Bluetooth: ISO: extend API for setting SDU interval

kruithofa opened the pull request:
The BT Core Spec v5.4 allows separate SDU_Interval to be set on C_To_P and P_To_C directions, but this is not possible with the existing interface. This PR splits the interval parameter in the call to bt_iso_sig_create into one for C_To_P and one for P_To_C

The same is done for the latency parameter

The babblesim tests are also updated


Fixes #62216

Commit:
Bluetooth: ISO: extend API for setting SDU interval

The BT Core Spec v5.4 allows separate SDU_Interval to
be set on C_To_P and P_To_C directions,
but this is not possible with the existing interface.

This PR splits the interval parameter in the call to
bt_iso_sig_create into one for C_To_P
and one for P_To_C

It also splits the latency parameter into one for
C_To_P and one for P_To_C

Signed-off-by: Andries Kruithof <andries.kruithof@nordicsemi.no>

Bluetooth: ISO: update UI for extended API

The API for setting the SDU interval and latency have been updated.
This PR also updates the setting of these by the user in the shell
and the iso_connected_benchmark sample

Signed-off-by: Andries Kruithof <andries.kruithof@nordicsemi.no>

Thalley commented on samples/bluetooth/iso_connected_benchmark/src/main.c:
```diff
@@ -924,7 +936,7 @@ static int change_central_settings(void)
 	printk("Change CIG settings (y/N)? (Current settings: cis_count=%u, "
 	       "interval=%u, latency=%u)\n",
-	       cig_create_param.num_cis, cig_create_param.interval,
+	       cig_create_param.num_cis, cig_create_param.c_to_p_interval,
```
While it will currently always be the same, suggest to log both anyways

kruithofa replied:
done

Thalley commented on samples/bluetooth/iso_connected_benchmark/src/main.c:
```diff
@@ -935,7 +947,7 @@ static int change_central_settings(void)
 
 		printk("New settings: cis_count=%u, inteval=%u, latency=%u\n",
-		       cig_create_param.num_cis, cig_create_param.interval,
+		       cig_create_param.num_cis, cig_create_param.c_to_p_interval,
```
While it will currently always be the same, suggest to log both anyways

Thalley commented on subsys/bluetooth/audio/bap_unicast_client.c:
```diff
@@ -2089,7 +2089,8 @@ static void bt_audio_codec_qos_to_cig_param(struct bt_iso_cig_param *cig_param,
 	cig_param->framing = qos->framing;
 	cig_param->packing = BT_ISO_PACKING_SEQUENTIAL; /*  TODO: Add to QoS struct */
-	cig_param->interval = qos->interval;
+	cig_param->c_to_p_interval = qos->interval;
+	cig_param->p_to_c_interval = qos->interval;
```
Hmm, we may need to support individual qos->interval values as well for the BAP API, but that's a separate fix. Ditto with latency.

kruithofa replied:
This one I'd like to put in a separate PR

Thalley commented on tests/bsim/bluetooth/host/iso/cis/src/cis_central.c:
```diff
@@ -237,7 +240,8 @@ static void reconfigure_cig(void)
 	param.cis_channels = &channels[1];
 	param.num_cis = 1U;
-	param.interval = interval_us; /* us */
+	param.c_to_p_interval = interval_us; /* us */
+	param.p_to_c_interval = interval_us; /* us */
```
We could consider adding a test that has different interval for the two if supported (ping @cvinayak is it?)

kruithofa replied:
yes, I am working on it

kruithofa replied:
done

kruithofa commented on include/zephyr/bluetooth/iso.h:
I'll add it to this PR

kruithofa commented on samples/bluetooth/iso_connected_benchmark/src/main.c:
```diff
@@ -854,7 +862,11 @@ static int parse_cig_args(void)
 #endif /* CONFIG_BT_ISO_ADVANCED */
 
-	cig_create_param.interval = interval;
+	if (role == ROLE_CENTRAL) {
+		cig_create_param.c_to_p_interval = interval;
+	} else {
+		cig_create_param.p_to_c_interval = interval;
+	}
```
I'll do that in a separate PR

kruithofa commented on samples/bluetooth/iso_connected_benchmark/src/main.c:
```diff
@@ -541,8 +546,11 @@ static int parse_interval_arg(void)
-	printk("Set interval (us) (current %u, default %u)\n",
-	       cig_create_param.interval, DEFAULT_CIS_INTERVAL_US);
+	interval = (role == ROLE_CENTRAL) ?
+		cig_create_param.c_to_p_interval : cig_create_param.p_to_c_interval;
```
done

Thalley commented on include/zephyr/bluetooth/iso.h:
```diff
@@ -335,19 +335,33 @@ struct bt_iso_cig_param {
 	 *  This value is ignored if any advanced ISO parameters are set.
 	 */
-	uint16_t latency;
+	uint16_t c_to_p_latency;
+
+	/** @brief Channel Latency in ms for SDUs sent from Peripheral to Central
+	 *
+	 *  Value range BT_ISO_LATENCY_MIN - BT_ISO_LATENCY_MAX.
+	 *
+	 *  This value is ignored if any advanced ISO parameters are set.
+	 */
+	uint16_t p_to_c_latency;
```
Since you've added latency to the same commit as interval, you'll either need to split that, or update the commit message and title

kruithofa replied:
commit message updated

kruithofa replied:
updated the commit message

Thalley commented on samples/bluetooth/central_iso/src/main.c:
```diff
@@ -251,8 +251,10 @@ int main(void)
 	param.framing = 0;
-	param.latency = 10; /* ms */
-	param.interval = interval_us; /* us */
+	param.c_to_p_latency = 10; /* ms */
+	param.p_to_c_latency = 10; /* ms */
```
Would probably make to have a value similar to `interval_us` we can use for both here, to avoid duplicate literal values

Thalley commented on subsys/bluetooth/host/iso.c:
```diff
@@ -1859,16 +1863,28 @@ static bool valid_cig_param(const struct bt_iso_cig_param *param, bool advanced,
-		LOG_DBG("Invalid interval: %u", param->interval);
+	if (param->c_to_p_interval < BT_ISO_SDU_INTERVAL_MIN ||
+	    param->c_to_p_interval > BT_ISO_SDU_INTERVAL_MAX) {
+		LOG_DBG("Invalid interval: %u", param->c_to_p_interval);
```
```suggestion
		LOG_DBG("Invalid C to P interval: %u", param->c_to_p_interval);
```

Thalley commented on subsys/bluetooth/host/iso.c:
```diff
@@ -1859,16 +1863,28 @@ static bool valid_cig_param(const struct bt_iso_cig_param *param, bool advanced,
+
+	if (param->p_to_c_interval < BT_ISO_SDU_INTERVAL_MIN ||
+	    param->p_to_c_interval > BT_ISO_SDU_INTERVAL_MAX) {
+		LOG_DBG("Invalid interval: %u", param->p_to_c_interval);
```
```suggestion
		LOG_DBG("Invalid P to C interval: %u", param->p_to_c_interval);
```

Thalley commented on subsys/bluetooth/host/iso.c:
```diff
@@ -1859,16 +1863,28 @@ static bool valid_cig_param(const struct bt_iso_cig_param *param, bool advanced,
+	if (!advanced &&
+	    (param->p_to_c_latency < BT_ISO_LATENCY_MIN ||
+	     param->p_to_c_latency > BT_ISO_LATENCY_MAX)) {
+		LOG_DBG("Invalid latency: %u", param->p_to_c_latency);
```
```suggestion
		LOG_DBG("Invalid P to C latency: %u", param->p_to_c_latency);
```

Thalley commented on subsys/bluetooth/host/iso.c:
```diff
@@ -1859,16 +1863,28 @@ static bool valid_cig_param(const struct bt_iso_cig_param *param, bool advanced,
-	if (param->interval < BT_ISO_SDU_INTERVAL_MIN ||
-	    param->interval > BT_ISO_SDU_INTERVAL_MAX) {
-		LOG_DBG("Invalid interval: %u", param->interval);
+	if (param->c_to_p_interval < BT_ISO_SDU_INTERVAL_MIN ||
+	    param->c_to_p_interval > BT_ISO_SDU_INTERVAL_MAX) {
```
Could also modify these to use `IN_RANGE` while we are modifying them now

kruithofa replied:
done

Thalley commented on subsys/bluetooth/shell/iso.c:
```diff
@@ -181,11 +181,17 @@ static int cmd_cig_create(const struct shell *sh, size_t argc, char *argv[])
+		param.p_to_c_interval = interval;
 	} else {
-		param.interval = 10000;
+		param.c_to_p_interval = 10000;
+		param.p_to_c_interval = 10000;
```
Suggest to avoid duplicating literal values

kruithofa replied:
yes, working on a commit that sets the values separately, which fixes this

Thalley commented on subsys/bluetooth/shell/iso.c:
```diff
@@ -247,9 +253,11 @@ static int cmd_cig_create(const struct shell *sh, size_t argc, char *argv[])
+		param.p_to_c_latency = latency;
 	} else {
-		param.latency = 10;
+		param.c_to_p_latency = 10;
+		param.p_to_c_latency = 10;
```
Suggest to avoid duplicating literal values

kruithofa replied:
agreed

Thalley commented on tests/bsim/bluetooth/host/iso/cis/src/cis_central.c:
```diff
@@ -234,10 +239,24 @@ static void reconfigure_cig(void)
 		return;
 	}
 
+	/* Test modifying to different values for both intervals */
+	param.num_cis = 0U;
+	param.c_to_p_interval = 5000U; /* us */
+	param.p_to_c_interval = 2500U; /* us */
+
+	err = bt_iso_cig_reconfigure(cig, &param);
+	if (err != 0) {
+		FAIL("Failed to reconfigure CIG to different intervals (%d)\n", err);
+
+		return;
+	}
```
Should we have something similar for different values of latency?

kruithofa replied:
yes, I don't think we are checking changing latencies at the moment but I can add it

kruithofa replied:
Come to think about it, arguably we should do the same for the other parameters in CIG create

Thalley replied:
> Come to think about it, arguably we should do the same for the other parameters in CIG create

Yeah, we probably want to improve ISO host test coverage :) 

The GMAP and CAP tests are doing quite a few different things, but doesn't hurt to have tests running at the ISO layer to also verify things like that

kruithofa replied:
for now just added the test for reconfiguring latency

Commit:
Bluetooth: ISO: use IN_RANGE for CIG create parameters

Use IN_RANGE instead of explicitly checking that a value is
in between two other values

Signed-off-by: Andries Kruithof <andries.kruithof@nordicsemi.no>

MariuszSkamra commented on subsys/bluetooth/shell/iso.c:
```diff
@@ -181,11 +181,17 @@ static int cmd_cig_create(const struct shell *sh, size_t argc, char *argv[])
+	 * setting it to the lowest interval number ensures that it
+	 * is increased at least once per interval
+	 */
+	cis_sdu_interval_us = MIN(param.c_to_p_interval, param.p_to_c_interval);
```
Not sure I follow this one. Shouldn't this be set to one of those depending on the role?
The `cis_sdu_interval_us` is later used to calculate the TX sequence number in 511.

```
	cis_sn_last = get_next_sn(cis_sn_last, &cis_sn_last_updated_ticks,
				  cis_sdu_interval_us);
```

Thalley replied:
Ah, nice catch. Given that it's used to generate the sequence number, it must be `c_to_p_interval` for central, and  `p_to_c_interval` for peripheral

Thalley commented on subsys/bluetooth/shell/iso.c:
```diff
@@ -136,6 +136,54 @@ NET_BUF_POOL_FIXED_DEFINE(tx_pool, 1, BT_ISO_SDU_BUF_SIZE(CONFIG_BT_ISO_TX_MTU),
+		shell_error(sh, "Invalid latency %lu", latency);
+
+		return -ENOEXEC;
+	}
+	return latency;
+}
+
+
+
```
```suggestion
	}
	
	return latency;
}

```

Thalley commented on subsys/bluetooth/shell/iso.c:
```diff
@@ -164,33 +215,48 @@ static int cmd_cig_create(const struct shell *sh, size_t argc, char *argv[])
+	 * CIS create error, so we do not need to set this variable.
+	 */
+	err = bt_conn_get_info(default_conn, &info);
+	if (err != 0) {
+		shell_error(ctx_shell, "Failed to get connection information: %d", err);
+	} else {
```
```suggestion
	if (err != 0) {
		shell_error(ctx_shell, "Failed to get connection information: %d", err);
		return -NOEXEC
	} else {
```

Thalley commented on subsys/bluetooth/shell/iso.c:
```diff
@@ -164,33 +215,48 @@ static int cmd_cig_create(const struct shell *sh, size_t argc, char *argv[])
+	if (err != 0) {
+		shell_error(ctx_shell, "Failed to get connection information: %d", err);
+	} else {
+		cis_sdu_interval_us = (info.role == BT_CONN_ROLE_CENTRAL) ?
+	  		              param.c_to_p_interval : param.p_to_c_interval;
```
Actually in `cmd_cig_create` we will always be the central (it's not allowed to create a CIG as the peripheral)

Maybe the role check should be put further up, and return early if not the central? 
Then this can just be `cis_sdu_interval_us = param.c_to_p_interval`.

The requirement to be a central in `HCI_LE_Set_CIG_Parameters` is arguably kind of weird, as the CIG can be created before any ACL connections...
